### PR TITLE
client: always mark exited sys/svc allocs as failed

### PR DESCRIPTION
**Pull request by schmichael**
When restarts.attempts=0 was set in a jobspec a system or service alloc
that exited with 0 status would be marked as `completed` instead of
`failed`. Since system and service jobs are intended to run until
stopped or updated, they should always be marked as failed when they
exit even in cases where the exit code is 0.

### CODE REVIEW

File: client/restarts.go [hunk 1]
```diff
@@ -127,7 +127,7 @@ func (r *RestartTracker) GetState() (string, time.Duration) {
 	// Hot path if no attempts are expected
 	if r.policy.Attempts == 0 {
 		r.reason = ReasonNoRestartsAllowed
-		if r.waitRes != nil && r.waitRes.Successful() {
+		if !r.onSuccess && r.waitRes != nil && r.waitRes.Successful() {
```
Comment (chelseakomlo) [inline client/restarts.go]: Maybe add a comment to explain this case.

File: client/restarts_test.go [hunk 2]
```diff
@@ -88,9 +88,25 @@ func TestClient_RestartTracker_ZeroAttempts(t *testing.T) {
...
 	rt := newRestartTracker(p, structs.JobTypeService)
 	if state, when := rt.SetWaitResult(testWaitResult(1)).GetState(); state != structs.TaskNotRestarting {
-		t.Fatalf("expect no restart, got restart/delay: %v", when)
+		t.Fatalf("expect no restart, got restart/delay: %v/%v", state, when)
+	}
+
+	// Even with a zero (successful) exit code non-batch jobs should exit
+	// with TaskNotRestarting
+	rt = newRestartTracker(p, structs.JobTypeService)
+	if state, when := rt.SetWaitResult(testWaitResult(0)).GetState(); state != structs.TaskNotRestarting {
```
Comment (chelseakomlo) [inline client/restarts_test.go]: Think about using require library.

File: client/restarts_test.go [hunk 3]
```diff
@@ -88,9 +88,25 @@ func TestClient_RestartTracker_ZeroAttempts(t *testing.T) {
...
+		t.Fatalf("expect no restart, got restart/delay: %v/%v", state, when)
+	}
+
+	// Batch jobs with a zero exit code and 0 attempts *may* exit cleanly
```
Comment (chelseakomlo) [inline client/restarts_test.go]: What is the difference between when they will and won't exit with this status?

Comment (schmichael) [reply client/restarts_test.go]: I'll improve the comment and add a test case for the batch+non-zero exit code case.

File: client/restarts.go [hunk 4]
```diff
@@ -127,10 +127,15 @@ func (r *RestartTracker) GetState() (string, time.Duration) {
 	// Hot path if no attempts are expected
 	if r.policy.Attempts == 0 {
 		r.reason = ReasonNoRestartsAllowed
-		if r.waitRes != nil && r.waitRes.Successful() {
+
+		// If the task does not restart on a successful exit code and
+		// the exit code was successful: terminate.
+		if !r.onSuccess && r.waitRes != nil && r.waitRes.Successful() {
```
Comment (preetapan) [inline client/restarts.go]: At this layer it doesn't seem to care about whether this is a batch or service job, is that something it should check for? e.g what about a batch job that has no restarts configured and it exited zero, in that case it should be marked complete and not failed..

Comment (schmichael) [reply client/restarts.go]: `onSuccess = !batch` which is super awkward and confuses me every time I look at this code so maybe I should just rename it to `isNotBatch`? or invert it to `isBatch`?

My default is to leave it be because I'm not sure if renaming would help the confusion. Let me know if you think a different name would be better!

**Comment by schmichael** [comment]
> Is there any debug-level logging or more logging in general that would be helpful here? -- @chelseakomlo 

I don't think so actually. We log the result of this method and include information about its decision in TaskEvents. I think what made it difficult to debug is that the method is fairly subtle/complex and has some poor internal naming.
